[자동차 경주 구현] 다이스 미션 제출합니다.

moondev03 opened the pull request:
### [궁금한 점]

```kotlin
fun generateCars(rawCars: String) {
    racingCars = rawCars.split(DELIMITER).map { Car(it.trim()) }

    require(racingCars.size > RACING_CAR_MIN_SIZE) { INVALID_CAR_SIZE_MESSAGE }
    require(racingCars.size == racingCars.map { it.getName() }.distinct().size) { DUPLICATE_CAR_NAME_MESSAGE }
}
```

- 위의 코드에서 입력값을 받아서 처리할 때, 다른 곳에서 받아서 처리하는 게 더 좋은 코드인지 궁금합니다.
- require()를 사용했는데, 이 부분을 다른 함수로 빼서 검증을 하는 것이 더 좋은 코드인지 궁금합니다.

<br>

```kotlin
// (1)
fun printResult(result: String) {
   println(result)
}

// (2)
fun printWinner(winner: String) {
    println("최종 우승자 : $winner")
}
```

- "최종 우승자 : " 부분을 winner 문자열에 포함해서 출력하면 하나의 함수로 합쳐서 구현할 수 있을 것 같은데, (1)과 (2) 코드 중에서 어떤 코드로 사용하는 게 더 적합한 코드인지 궁금합니다.

<br>

```kotlin
@ParameterizedTest
@ValueSource(strings = ["1", "22", "333", "4444", "55 5 "])
fun `자동차 이름 유효성 테스트`(carName: String) {
    assertDoesNotThrow { Car(carName) }
}
```

- 테스트 코드에 source를 여러개 넣어서 테스트하고 있는데, 테스트케이스명을 명시적으로 해서 테스트 하는 게 좋은지 궁금합니다.

gahyunkim committed:
docs(readme): 기능 요구사항 정의

gahyunkim committed:
feat: 경주할 자동차 이름 입력 받기

- 자동차 생성
- 자동차 이름 예외 처리
- 테스트 코드 작성

gahyunkim committed:
feat: 경주 실행 기능 구현

- 시도할 횟수 입력 기능 구현
- 경주 실행 기능 구현
- 예외 처리
- 테스트 코드 작성

gahyunkim committed:
feat: 경주 결과 출력 기능 구현

- 경주 현황 출력 구현
- 경주 우승자 결정 기능 구현

gahyunkim committed:
feat: Car 테스트 코드 구현

- 자동차 생성 및 예외 테스트 코드 구현

gahyunkim committed:
refactor: 경주 횟수 검증 로직 수정

gahyunkim committed:
chore: Ktlint 수정

gahyunkim committed:
refactor: tryRacing 함수의 문자열 연결 방식 최적화

gahyunkim committed:
refactor: calculateWinner() 코드 최적화

- maxOfOrNull -> maxOf
- map, joinToString merge

gahyunkim committed:
refactor: RacingGame 상수화 적용

gahyunkim committed:
refactor: 랜덤 번호 생성 로직 분리

- 테스트 코드 기능 구현
- 테스트 코드 이름 수정

gahyunkim committed:
refactor: 상수 이름 수정

gahyunkim committed:
feat: 경주 종료 후 복수 우승자 검토 테스트 코드 구현

gahyunkim committed:
refactor: WINNER_SEPARATOR 상수화

KwonDae commented on README.md:
README를 잘 작성해주셨습니다 👍 
하나 생각해볼점은, 기능 목록 단위를 커밋단위로 추가하게 된다면(README 에서 작성한 기능목록 단위와 커밋단위가 일치)
어떤 이점이 있을지 생각해볼까요?

moondev03 replied:
제이든, 안녕하세요!
좋은 리뷰 감사합니다 :)

기능 목록 단위와 커밋 단위가 일치한다면, 기능별로 어떤 변경이 이루어졌는지 파악하기 비교적 수월해지는 이점이 있을 것 같아요!
앞으로는 최대한 일치되도록 해보겠습니다!

KwonDae commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -0,0 +1,17 @@
+package view
+
+class InputView {
```
Input과 Output을 잘 분리해주셨네요.
class를 활용할 수 있지만, object도 활용해볼 수 있을거 같은데
어떤 차이점이 있을까요?

moondev03 replied:
해당 미션에서는 아래와 같은 이유 때문에 Object를 활용하는 것이 더 적절했던 것 같아요!

1. View에서 상태를 가질 필요가 없다.
2. 여러 인스턴스를 필요로 하지 않는다.

KwonDae commented on src/main/kotlin/model/RacingGame.kt:
```diff
@@ -0,0 +1,49 @@
+    fun generateCars(rawCars: String) {
+        racingCars = rawCars.split(DELIMITER).map { Car(it.trim()) }
+
+        require(racingCars.size > RACING_CAR_MIN_SIZE) { INVALID_CAR_SIZE_MESSAGE }
+        require(racingCars.size == racingCars.map { it.getName() }.distinct().size) { DUPLICATE_CAR_NAME_MESSAGE }
```
require 활용 👍

KwonDae replied:
다만, 입력값을 검증하는 책임은 어느 클래스에게 가는게 좋을까요?
현재는 InputView에 역할을 주고, RacingGame에서는 InputView에서의 결과값에 따라 판단을 할 수 있을거 같은데요.

클래스를 잘 분리해주신 만큼, 각 클래스(객체)에게 메시지를 던져보는건(입력을 받아!) 어떨까요?

KwonDae replied:
> 위의 코드에서 입력값을 받아서 처리할 때, 다른 곳에서 받아서 처리하는 게 더 좋은 코드인지 궁금합니다.

위 내용과 비슷합니다. 입력을 받았을 때 먼저 검증을 하도록 Input쪽에 역할을 부여해보는건 어떨까요?
테스트 관점에서 어떤 차이가 있을지 고민해보시고, **저에게도 공유해주세요!**

> require()를 사용했는데, 이 부분을 다른 함수로 빼서 검증을 하는 것이 더 좋은 코드인지 궁금합니다.

이부분은 위 내용대로 각 객체에게 책임을 부여하다보면 어떻게 검증을 해야 좋을지 고민하시게 될 것 같아 직접 구현해보시고 다음 리뷰때 생각을 나눠보시죠!

moondev03 replied:
말씀해주신 것처럼 `RacingGame`에서 위치하던 입력값 검증 책임을 `InputView`로 변경하니, 
이제 입력값 검증 테스트를 위해 `RacingGame` 객체를 생성할 필요 없이 `InputView`만으로 바로 테스트할 수 있게 되었습니다.

입력값에 대한 책임을 어디에 둬야 할 지 고민을 계속 했었는데, 
이제는 `InputView`에 책임을 지도록 하는게 더 좋은 방법인 것 같다고 느껴지네요 :)

KwonDae commented on src/test/kotlin/model/RacingGameTest.kt:
```diff
@@ -0,0 +1,68 @@
+        val racingGame = RacingGame(generator)
+        racingGame.generateCars("carA, carB")
+        assertThrows<IllegalArgumentException> { racingGame.tryRacing(input) }
+    }
```
입려값에 대한 테스트는 RacingGame에 대한 테스트일까요?

moondev03 replied:
입력값 검증 책임을 `RacingGame`에서 `InputView`로 변경함으로써

```kotlin
val generator = RandomNumberGeneratorImpl()
val racingGame = RacingGame(generator)
```
위 객체를 생성할 필요없이 `InputView.inputXXX`를 통해 바로 테스트가 가능해졌습니다.

또한 입력값 검증 책임의 위치 변경으로 인해 `InputViewTest` 파일을 만들어 위치하도록 했습니다!

KwonDae commented on src/main/kotlin/model/RandomNumberGenerator.kt:
```diff
@@ -0,0 +1,8 @@
+package model
+
+interface RandomNumberGenerator {
```
인터페이스를 활용한 전략패턴 좋습니다 💯

KwonDae commented on src/test/kotlin/model/CarTest.kt:
```diff
@@ -0,0 +1,45 @@
+    @ValueSource(strings = ["", " ", "33333", "44 4444"])
+    fun `자동차 이름 유효성 예외 테스트`(carName: String) {
+        assertThrows<IllegalArgumentException> { Car(carName) }
+    }
```
> 각 자동차에 이름을 부여할 수 있다. 자동차 이름은 5자를 초과할 수 없다.

요구사항중에 맞지 않는게 하나 있네요!

moondev03 replied:
앗, 제대로 확인하지 못했었네요 😭
확인 감사합니다!

KwonDae commented on src/main/kotlin/model/Car.kt:
```diff
@@ -0,0 +1,25 @@
+package model
+
+class Car(private val name: String) {
+    private var step: Int = DEFAULT_STEP
```
💯

KwonDae commented on src/main/kotlin/view/OutputView.kt:
```diff
@@ -0,0 +1,11 @@
+package view
+
+class OutputView {
+    fun printResult(result: String) {
+        println(result)
+    }
+
+    fun printWinner(winner: String) {
+        println("최종 우승자 : $winner")
+    }
```
> "최종 우승자 : " 부분을 winner 문자열에 포함해서 출력하면 하나의 함수로 합쳐서 구현할 수 있을 것 같은데, (1)과 (2) 코드 중에서 어떤 코드로 사용하는 게 더 적합한 코드인지 궁금합니다.

이부분은 정답은 없다고 생각합니다. 출력 기능 요구사항에 따라 달라질 것 같은데요,
승패, 결과 안내시 `printResult` 함수가 공통으로 자주 쓰인다면 지금 처럼 분리하는것도 좋습니다 👍

KwonDae commented on src/test/kotlin/model/RacingGameTest.kt:
```diff
@@ -0,0 +1,68 @@
+
+class RacingGameTest {
+    @ParameterizedTest
+    @ValueSource(strings = ["", " ", "car12345", "1", "car, car", "car1, car12345"])
```
> 테스트 코드에 source를 여러개 넣어서 테스트하고 있는데, 테스트케이스명을 명시적으로 해서 테스트 하는 게 좋은지 궁금합니다.

같은 테스트 항목에 대해 여러 케이스를 보여줄 수 있어 구현 해주신 것 처럼 source를 활용하는것도 좋습니다. 실제로 저 또한 사용을 하고 있습니다. 
확실한건 "명확한 테스트케이스명" + 여러 케이스 (source)를 적절히 함께 사용하면 코드를 보는 동료들도 보다 빠르고 명확하게 이해할 수 있을거에요

KwonDae reviewed (changes requested):
안녕하세요 다이스!
자동차 경주 첫 미션 함께 진행하게 된 제이든입니다 :) 

잘 구현해 주셔서 생각해 보면 좋을 만한 부분들에 코멘트를 남겨드렸어요.
천천히 읽어보시고 궁금한 부분이 있다면 언제든 DM 주시면 감사하겠습니다.

moondev03 committed:
fix: 자동차 이름 길이 조건 수정

- 최대 이름 길이 4글자 -> 5글자로 수정
- 테스트 케이스 수정

moondev03 committed:
refactor: Car name private 접근자 제거

- name 접근자를 public으로 수정
- getName() 제거

moondev03 committed:
chore: 패키지 구조 변경에 따른 파일 이동

- racingcar 패키지 아래에 model, view, controller 배치

moondev03 committed:
chore: NumberGenerator Interface 네이밍 수정

- RandomNumberGeneratorImpl -> RandomNumberGenerator 이름 수정
- TestNumberGenerator -> FixedNumberGenerator 이름 수정

moondev03 committed:
refactor: 입력 문자열 검증 위치 변경

- RacingGame에서 수행하던 자동차 이름과 경주 횟수 입력값 검증 로직을 InputView로 변경
- 테스트 코드 작성

moondev03 committed:
refactor: InputView & OutputView Object로 변경

moondev03 committed:
refactor: Car Step Custom Getter 제거 및 접근자 수정

- private -> public
- setter는 private로 제한

moondev03 committed:
refactor: RacingGame & Controller 리팩토링

- 입력값 검증 로직 제거
- Controller에서 RacingCar 미션의 수행 순서를 준비-경주 수행-결과로 정리
- tryRacing 모든 수행 과정을 한번에 문자열로 반환하는 대신 각 시도마다 전진 현황을 출력할 수 있도록 변경

moondev03 committed:
chore: gitkeep 파일 제거

moondev03 committed:
refactor: tryRacing() indent depth 제한 반영

moondev03 committed:
fix: 실행 결과 헤더 출력 순서 수정

moondev03 committed:
fix: 자동차 이름 빈문자열인 경우 필터링

moondev03 committed:
feat: 자동차 이름 & 경주 횟수 입력 검증 테스트 추가

- setConsoleInput() 호출 위치 수정

KwonDae commented on src/main/kotlin/racingcar/controller/RacingController.kt:
```diff
@@ -0,0 +1,43 @@
+import racingcar.view.InputView
+import racingcar.view.OutputView
+
+class RacingController {
+    private var racingGame: RacingGame
```
RacingGame을 외부로부터 받을 수 있을까요?
외부로부터 주입받으면 어떤 이점이 생길 수 있을까요?

moondev03 replied:
외부에서 받아오도록 하면 테스트 용이성이 좋아지고
Controller가 객체 생성의 책임으로부터 분리될 수 있을 것 같네요!

KwonDae commented on src/main/kotlin/racingcar/controller/RacingController.kt:
```diff
@@ -0,0 +1,43 @@
+    private var racingGame: RacingGame
+
+    init {
+        val randomNumberGeneratorImpl = RandomNumberGenerator()
```
위 내용과 비슷합니다.
내부에서 객체를 생성하는것과 외부에서 받는것 
어떤 차이가 있을까요?

KwonDae commented on src/main/kotlin/racingcar/controller/RacingController.kt:
```diff
@@ -0,0 +1,43 @@
+
+    init {
+        val randomNumberGeneratorImpl = RandomNumberGenerator()
+        racingGame = RacingGame(randomNumberGeneratorImpl)
```
또한, 초기화 단계에서 셋팅이 되는거라면 val 사용하여 초기화할 수 있습니다.

KwonDae reviewed (approved):
1단계 피드백 반영하시느라 고생하셨습니다.
잘 진행해주신 만큼 다음 단계에서도 더 깊게 이야기 해볼 수 있을거 같아요!
다음 단계 진행하면서 생각해보면 좋을 만한 부분에 코멘트를 남겼으니 참고해서 진행해주세요!
